lib/agent.js

```diff
@@ -0,0 +1,70 @@
+'use strict'
+
+const { Pool } = require('../')
+
+const kAgentPoolOpts = Symbol('agent pool opts')
+const kAgentCache = Symbol('agent cache')
+const kAgentRegistry = Symbol('agent registry')
+class Agent {
+    constructor (opts) {
+        this[kAgentPoolOpts] = opts
+        this[kAgentCache] = new Map()
+        this[kAgentRegistry] = new FinalizationRegistry(key => {
+            // get the Pool WeakRef from the cache
+            const ref = this[kAgentCache].get(key)
+            // if the WeakRef exists and the object has been reclaimed
+            // according to mdn the second logic operation here is superfluous as a deref call in a FR will always return undefined
+            if (ref !== undefined && ref.deref() === undefined) {
+                // remove the WeakRef from the cache
```

Comment:
Would be nice with an assertion checking that there are not pending requests in the pool

Reply:
what should this callback do in the case there are requests pending or in flight? Also any ideas how to test this scenario? As far as I understand it the GC would not clean up a Pool currently being used

Reply:
> what should this callback do in the case there are requests pending or in flight? 

It shouldn't be possible. If that happens it's a bug which is why I suggest an assertion.

Reply:
Actually never mind. It's not possible to assert this.